### DISCUSSION TITLE
[BP#20] Générer des spritesheets CSS

### DISCUSSION THREAD

**Discussion by florinesueur** [cnumr/best-practices#223] [Bonnes pratiques] [answered]
Il me semble qu'il faudrait un exemple un peu plus parlant où il est intéressant d'utiliser les spritessheets CSS.
Par exemple pour les petits drapeaux de changement de langue d'un site internet :
```
.flags {
  background-image: url('../images/flags.png');
  background-repeat: no-repeat;
}

.flag-france {
  height: 128px;
  background-position: -5px -5px;
}

.flag-usa {
  height: 135px;
  background-position: -5px -143px;
}

```
ou

`<img src='flags.png' class='flag-france' alt='Français'> `

```
img {
  object-fit: none;
  object-position: 0 0;
  width: 100px;
  height: 100px;
}

.flag-france {
object-position: -5px -5px;
}
```

Aussi nous pouvons proposer une alternative en lien avec la bonne pratique [BP30](https://github.com/cnumr/best-practices/blob/main/chapters/BP_030_fr.md) en faire des glyphes.

http://draeton.github.com/stitches est un lien mort.

**Comment by ldesgrange** [comment]
Je suis d'accord pour un exemple plus parlant.

Par contre, comme pour la [BP09](https://github.com/cnumr/best-practices/blob/main/chapters/BP_009_fr.md), HTTP/2 rend ce genre de techniques moins pertinentes et il faut faire des tests au cas par cas pour voir si on y gagne, donc à mon avis, il faudrait revoir à la baisse l'impact environnemental de cette BP.

**Comment by Fred-artwai** [comment, marked as answer]
Mais qui utilise encore des sprites CSS aujourd'hui ? ;-)
Pour moi, çà n'a pas de sens. Faire une sprite CSS c'est limiter le nombre de requêtes donc 
soit c'est compris dans BP09 comme une technique pour réduire le nombre de requête
soit c'est tout simplement à supprimer car en opposition avec : 
- la BP 30 Préférer les glyphs aux images
- la BP 37 Utiliser le chargement paresseux des images (très compliqué en CSS)

**Reply by arnolem** [reply]
Qui ?
Google par exemple : 
 - https://ssl.gstatic.com/gb/images/p1_4d827a84.png
 - https://www.google.com/images/searchbox/desktop_searchbox_sprites318_hr.webp
 
 Je suis d'accord avec toi mais je pense que cette pratique est encore d'actualité malgré tout. Et beaucoup de sites web ne sont pas encore en HTTP/2

**Reply by florinesueur** [reply]
Je ne suis pas forcément d'accord avec @Fred-artwai sur les deux points d'oppositions : 
- les colors font (glyphs à multicouches de couleurs) sont encore mal prient en charge avec les navigateurs [https://design.tutsplus.com/tutorials/what-are-color-fonts--cms-31467](https://design.tutsplus.com/tutorials/what-are-color-fonts--cms-31467) donc pas possible pour des drapeaux, par exemple.
- Le chargement paresseux est possible avec le deuxième exemple d'utilisation sur un élément img.

**Reply by florinesueur** [reply]
Je pense que nous pouvons supprimer cette BP en incluant comme alternative à la BP30 : Préférer les glyphs aux images et de mettre un disclaimer dans la nouvelle BP sur les Gifs animés en indiquant que ce n'est pas une solution valable de remplacement aux gifs animés.

**Reply by Fred-artwai** [reply]
Autant pour moi l'exemple 2 est excellent ! Mais je pense quand même qu'il faut supprimer cette BP et encourager
- moins de requêtes en HTTP/1, 
- l'HTTP/2 (voir le HTTP/3) 
- et les glyphes dans le cas d'un usage monochrome.

**Reply by ACTLEM** [reply]
Une BP HTTP2 est bien prévue (déjà écrite) pour cette version. Par contre, j'ai du mal à comprendre les arguments pour la suppression pour cette BP, quelle serait la meilleure alternative si nous ne sommes pas sur du monochrome?

**Reply by Fred-artwai** [reply]
A partir du moment où nous sommes en HTTP/2 cette pratique a beaucoup moins d'intérêts. Et dans tous les cas c'est un moyen de réduire le nombre de requêtes donc a citer comme exemple dans la BP#9.

**Reply by tbroyer** [reply]
Sans compter que le moindre changement d'un sprite (une image parmi toutes) nécessite de re-télécharger la spritesheet (l'assemblage des sprites), alors qu'avec des images distinctes on bénéficie du cache HTTP pour chacun indépendamment.
L'assemblage d'images variées a aussi un impact sur les optimisations possibles au niveau de l'image (plus grande palette de couleurs qui empêche l'utilisation de PNG-8 par exemple)

Pour moi, les spritesheets ont encore un intérêt dans certains cas, mais bien moins qu'il y a quelques années, et ne sont plus une règle de base. Je serais tenté de supprimer cette BP, mais je pense qu'il faudrait faire des mesures (ou trouver des études déjà réalisées sur le sujet)

**Reply by ACTLEM** [reply]
On pourrait garder la référence au Spritesheets dans la BP9: https://github.com/cnumr/best-practices/discussions/203 et supprimer cette BP, ok?

**Reply by ACTLEM** [reply]
J'ai créé l'issue pour suppression mais on garde une référence aux spritesheets dans la BP9: https://github.com/cnumr/best-practices/issues/299
